File: problems/923/solution.go
package main
const MOD int = 1000000007;

func threeSumMulti(arr []int, target int) int {
    dp := make([][]int, 3)
    
    for i := range dp {
        dp[i] = make([]int, target + 1)
    }
    
    for _, num := range arr {
        if num > target {
            continue
        }
        
        for i := target - num; i >= 0; i-- {
            dp[2][i + num] += dp[1][i]
            dp[2][i + num] %= MOD
            dp[1][i + num] += dp[0][i]
            dp[1][i + num] %= MOD
        }
        
        dp[0][num]++
    }
    
    return dp[2][target] % MOD
}


/*
arr = [1,1,2,2,3,3,4,4,5,5], target = 8
    1  2  3  4  5  6  7  8
0   2  2  2  2  2  0  0  0 
1   0  1  4  5  8  9  8  5 
2   0  0  0  2  4  10 14 20 

arr = [1,1,2,2,2,2], target = 5
    1  2  3  4  5
0   2  4  0  0  0 
1   0  1  8  6  0 
2   0  0  0  4  12 


*/

File: problems/923/solution.cpp
#define ll long long int

const int MOD = 1e9 + 7;

class Solution {
public:
    int threeSumMulti(vector<int>& arr, int target) {
        ll ans = 0;
        unordered_map<ll, ll> freq;
        
        for (const int& num: arr) {
            freq[num]++;
        }
        
        for (auto [i, ignore]: freq) {
            for (auto [j, ignore]: freq) {
                ll k = target - i - j;
                
                if (not freq.count(k)) {
                    continue;
                }
                
                if (i == j and j == k) {
                    ans += freq[i] * (freq[j] - 1) * (freq[k] - 2) / 6;
                } else if (i == j and j != k) {
                    ans += freq[i] * (freq[j] - 1) / 2 * freq[k];
                } else if (i < j and j < k) {
                    ans += freq[i] * freq[j] * freq[k];
                }
            }
        }
        
        return ans % MOD;
    }
};